### ISSUE TITLE
[dart2wasm] Consider using inferred-type TFA metadata to generate more precisely typed locals, functions, fields

### ISSUE THREAD

**Issue by osa1** [dart-lang/sdk#53931]
TFA's inferred type metadata maps `Field`s, call sites (`InstanceInvocation`, `InstanceGet`, `FunctionInvocation` etc. nodes), function arguments (`VariableDeclaration`) (and probably other nodes) to more precise types than their Dart types.

We should consider using this metadata to generate more precisely typed locals, function types, and fields.

**Comment by askeksa** [comment]
Good idea. If we do the parameters and fields, I expect Binaryen will take care of the remaining locals.

**Comment by osa1** [comment]
Closing as complete. We now use TFA-inferred types for arguments: https://github.com/dart-lang/sdk/blob/3873af28e9654857b4afdb5a074670589034746e/pkg/dart2wasm/lib/translator.dart#L1448

Locals and fields: https://github.com/dart-lang/sdk/blob/3873af28e9654857b4afdb5a074670589034746e/pkg/dart2wasm/lib/translator.dart#L1517-L1527

Return types: https://github.com/dart-lang/sdk/blob/3873af28e9654857b4afdb5a074670589034746e/pkg/dart2wasm/lib/translator.dart#L1484-L1488

**Comment by mkustermann** [comment]
> Locals and fields:

Actually I think TFA doesn't emit metadata for normal local variables - only in very limited circumstances.
